---
title: "expressyouRcell"
author: "Martina Paganin"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{expressyouRcell}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


# Overview

__expressyouRcell__ is a unique and easy-to-use R package which provides an intuitive approach for visualizing and presenting multi-dimensional variations of gene expression levels across time and space. This tool gives the possibility of generating animations of pictographic representations of cells, or pictograms, providing a convenient and intuitive method for visualizing and understanding time course variations in cellular compartments. A range of customizable options is provided to create cellular pictograms starting from users's data.

# Before starting

## Dependencies

* CRAN
	+ data.table (>= 1.13.6),
	+ ggplot2 (>= 3.3.3),
	+ rsvg (>= 2.1),
	+ grImport2 (>= 0.2),
	+ magick (>= 2.5.2),
	+ ggpubr (>= 0.4.0),
	+ RColorBrewer (>= 1.1-2)  
	 
* Bioconductor
	+ IRanges (>= 2.24.1),
	+ org.Mm.eg.db (>= 3.12.0),
	+ clusterProfiler (>= 3.18.0),  
	+ DOSE (>= 3.16.0),	
	+ multtest (>= 2.46.0)
	
## Installation

To install __expressyouRcell__ directly from GitHub, the *devtools* package is required. If not already installed on your system, run:

```{r, eval = FALSE}
install.packages("devtools")
```

Otherwise, load *devtools* and install __expressyouRcell__ by:

```{r, eval = FALSE}
library(devtools)
install_github("https://github.com/gittina/expressyouRcell", dependencies = TRUE)
```

# Loading

To load __expressyouRcell__ run:

```{r, eval = TRUE, warning = FALSE}
library(expressyouRcell)
```

# Prepare input data
__expressyouRcell__ is optimized for representing multiple sets of gene expression data (e.g. multiple stages). For this reason, the input has to be organized as a list of *data.tables*. For example, in case of multiple stages, each *data.table* should correspond to a specific time point.

Each *data.table* must have at least a column of gene names named precisely `gene_symbol`. The input table can also contain additional columns with values of gene expression levels (CPM or FPKM) or results from upstream differential analysis pipeline (such as fold changes and pvalues).

# Create the gene-localization table
Before computing the color shade of each region, genes have to be associated with their cellular localization within the cell structure. So, for this step a table with information on gene localization within cellular organelles is needed. A custom table with this information, if the user already has it, can be used by __expressyouRcell__ for creating the cellular pictograms. In this case, the table must contain two columns: one with gene names (named exactly `gene_symbol`) and one with the associated information of the localization of that gene within the cell (named exactly `subcell_struct`).

Otherwise, the user can create the gene-localization table with the `map_gene_localization` function provided within __expressyouRcell__

__expressyouRcell__ offers two options for associating each gene with its localization.
The first option requires as input the gene annotation file, in GTF format, used during the alignment step of the samples. The second one, requires as input a list of *data.tables* where data have been stored. It is mandatory to organize the input datasets as a list of *data.tables*, and each one must contain a column with names of the genes named precisely `gene_symbol`. On this complete set of gene symbols, a gene ontology enrichment analysis is performed to associate a gene with a term in the cellular component ontology. For this purpose, only the sub-ontology of the cellular components is taken into consideration. This function generates the gene-localization table, which maps each gene to the locations in the cellular structures, either cellular compartments or macromolecular complexes.

Example of usage with the annotation GTF file:

```{r, eval = FALSE}
gene_loc_table <- map_gene_localization(gene_set = paste0(utils_folder, "gencode.vM22.primary_assembly.annotation.gtf"))
```

Example of usage with the list of *data.table* provided with the package:

```{r, eval = TRUE, warning = FALSE}
gene_loc_table <- map_gene_localization(gene_set = example_list)
```

# Print and save your results 
# Create the gene-localization table

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css
        
You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
