---
title: "expressyouRcell"
author: "Martina Paganin"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{expressyouRcell}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


# Overview

__expressyouRcell__ is a unique and easy-to-use R package which provides an intuitive approach for visualizing and presenting multi-dimensional variations of gene expression levels across time and space. This tool gives the possibility of generating animations of pictographic representations of cells, or pictograms, providing a convenient and intuitive method for visualizing and understanding time course variations in cellular compartments. A range of customizable options is provided to create cellular pictograms starting from users's data.

# Before starting

## Dependencies

* CRAN
	+ data.table (>= 1.13.6),
	+ ggplot2 (>= 3.3.3),
	+ rsvg (>= 2.1),
	+ grImport2 (>= 0.2),
	+ magick (>= 2.5.2),
	+ ggpubr (>= 0.4.0),
	+ RColorBrewer (>= 1.1-2)  
	 
* Bioconductor
	+ IRanges (>= 2.24.1),
	+ org.Mm.eg.db (>= 3.12.0),
	+ clusterProfiler (>= 3.18.0),  
	+ DOSE (>= 3.16.0),	
	+ multtest (>= 2.46.0)
	
## Installation

To install __expressyouRcell__ directly from GitHub, the *devtools* package is required. If not already installed on your system, run:

```{r, eval = FALSE}
install.packages("devtools")
```

Otherwise, load *devtools* and install __expressyouRcell__ by:

```{r, eval = FALSE}
library(devtools)
install_github("https://github.com/gittina/expressyouRcell", dependencies = TRUE)
```

# Loading

To load __expressyouRcell__ run:

```{r, eval = TRUE, warning = FALSE}
library(expressyouRcell)
```

# Prepare input data
__expressyouRcell__ is optimized for representing multiple sets of gene expression data (e.g. multiple stages). For this reason, the input has to be organized as a list of *data.tables*. For example, in case of multiple stages, each *data.table* should correspond to a specific time point.

Each *data.table* must have at least a column of gene names named precisely `gene_symbol`. The input table can also contain additional columns with values of gene expression levels (CPM or FPKM) or results from upstream differential analysis pipeline (such as fold changes and pvalues).

# Create the gene-localization table
Before computing the color shade of each region, genes have to be associated with their cellular localization within the cell structure. So, for this step a table with information on gene localization within cellular organelles is needed. A custom table with this information, if the user already has it, can be used by __expressyouRcell__ for creating the cellular pictograms. In this case, the table must contain two columns: one with gene names (named exactly `gene_symbol`) and one with the associated information of the localization of that gene within the cell (named exactly `subcell_struct`).

Otherwise, the user can create the gene-localization table with the `map_gene_localization` function provided within __expressyouRcell__

__expressyouRcell__ offers two options for associating each gene with its localization.
The first option requires as input the gene annotation file, in GTF format, used during the alignment step of the samples. The second one, requires as input a list of *data.tables* where data have been stored. It is mandatory to organize the input datasets as a list of *data.tables*, and each one must contain a column with names of the genes named precisely `gene_symbol`. On this complete set of gene symbols, a gene ontology enrichment analysis is performed to associate a gene with a term in the cellular component ontology. For this purpose, only the sub-ontology of the cellular components is taken into consideration. This function generates the gene-localization table, which maps each gene to the locations in the cellular structures, either cellular compartments or macromolecular complexes.

Example of usage with the annotation GTF file:

```{r, eval = FALSE}
gene_loc_table <- map_gene_localization(gene_set = paste0(utils_folder, "gencode.vM22.primary_assembly.annotation.gtf"))
```

Example of usage with the list of *data.table* provided with the package:

```{r, eval=TRUE, message=FALSE, warning=FALSE}
gene_loc_table <- map_gene_localization(gene_set = example_list)
```

# Choose and color the cellular pictogram

# Visualize pictogram with default colors
The `plot_cell` function allows the user to visualize the chosen cellular map with the default colors. It requires as input the *data.table* with the graphical information (coordinates and colors for the cellular organelles). 

```{r eval=TRUE, fig.height=5, fig.width=10, message=FALSE, warning=FALSE}
plot_cell(coords_dt = "neuron")
```

## Visualize user's data on the cellular pictogram
The main function is called `color_cell` and needs at least three mandatory input parameters.
* A list of *data.tables*, each corresponding to a time point and must have at least a column named `external_gene_name` with gene symbols.
* A *data.table* containing `x` and `y` coordinates, name of the subcellular structures and an associated default color.
* A *data.table* storing for each gene the mapping to a subcellular localization according to the cellular component gene ontology. 

Different methods for assigning colors to subcellular localizations can be chosen through the `coloring_method` parameter. 

### Mean (or median) of values
If  `coloring_method` is equal to `mean` or `median`,  genes are first grouped according to their localization, then, mean (or median) of numeric values associated with each gene is computed for each group. In this case, the *data.tables*  in the input list must also have an additional column containing numeric values (e.g. logFC, CPM values, etc.). To specify the value column on which you want to base your coloring, an additional parameter with the name of the column (`col_name`) must be provided as input. The given name must be compatible with column names in the *data.tables* of the input list. 

expressyouRcell can handle your output in two main different manners, and this can be achieved with the optional parameters in the ```color_cell``` function.

#### 1) Classify genes into separate groups and for each one generate a distinct plot
expressyouRcell allows you to selectively visualize only genes belonging to distinct classes (e.g. either "up-" or "down-regulated" genes) and generating separate plots for each of the specified categories of genes. In this case, separate analysis for each subset of genes can be performed, and expressyouRcell will then output single ```ggplot``` objects for each category. 
To select this analysis, you must specify as the ```group_by``` parameter the name of a column with the categorical variable (e.g. “class”) on which you have previuoly stored the gene classification. The additional parameter ```grouping_vars``` can be specified to subselect the categories you are interested to plot (e.g. in case of DEGs classification, “up” and “down”). Default value of this parameter is null. In this case, all the genes are selected and their corresponding values are averaged for each subcellular localization, regardless any classification.

```{r eval=TRUE, fig.height=5, fig.width=10, message=FALSE, warning=FALSE}
example_list_output <- color_cell(timepoint_list = example_list,
                                  plot_data = "neuron",
                                  gene_loc_table = gene_loc_table,
                                  coloring_mode = "mean",
                                  group_by = "class",
                                  grouping_vars = list("class"=c("+","-")),
                                  col_name = "logFC",
                                  colors = list("+" = c("#eaf3ea", "#307e2d"),
                                                "-" = c("#f3eaea", "#7e302d")))
```

# Print and save your results 


